File: apps/social/src/lib/parse-public-blog-index.ts
/**
 * Parse public blog listing markdown (`blog/index.md`) into post links.
 * Matches lines produced by `buildBlogIndexMarkdown` in `@envoymesh/api`.
 */
export interface PublicBlogPostLink {
  title: string;
  url: string;
}

const LINK_RE = /^-\s*\[([^\]]+)\]\((envoy:\/\/[^)\s]+)\)/gm;

export function parsePublicBlogIndexMarkdown(markdown: string): PublicBlogPostLink[] {
  const text = markdown.trim();
  if (!text || text.includes("_No posts yet._")) return [];
  const out: PublicBlogPostLink[] = [];
  for (const m of text.matchAll(LINK_RE)) {
    const title = m[1]?.trim();
    const url = m[2]?.trim();
    if (!title || !url) continue;
    out.push({ title, url });
  }
  return out;
}

File: apps/envoygo/lib/services/parse_public_blog_index.dart
/// Parse public blog listing markdown (`blog/index.md`) into post links.
final _blogIndexLinkRe =
    RegExp(r'^-\s*\[([^\]]+)\]\((envoy://[^)\s]+)\)', multiLine: true);

class PublicBlogPostLink {
  final String title;
  final String url;
  const PublicBlogPostLink({required this.title, required this.url});
}

List<PublicBlogPostLink> parsePublicBlogIndex(String markdown) {
  final text = markdown.trim();
  if (text.isEmpty || text.contains('_No posts yet._')) return const [];
  return _blogIndexLinkRe
      .allMatches(text)
      .map(
        (m) => PublicBlogPostLink(
          title: m.group(1)!.trim(),
          url: m.group(2)!.trim(),
        ),
      )
      .where((e) => e.title.isNotEmpty && e.url.isNotEmpty)
      .toList();
}
